src/SqlAsyncCollector.cs

```diff
@@ -261,32 +295,40 @@ private static void GenerateDataQueryForMerge(TableInformation table, IEnumerabl
-
-                // Look up primary key of T. Because we're going in the same order of fields every time,
-                // we can assume that if two rows with the same primary key are in the list, they will collide
-                foreach (PropertyInfo primaryKey in table.PrimaryKeys)
+                // When T is JObject, currently there isn't way to find out the Primary keys, so it'll be null, adding check.
+                if (table.PrimaryKeys.Any())
                 {
-                    object value = primaryKey.GetValue(row);
-                    // Identity columns are allowed to be optional, so just skip the key if it doesn't exist
-                    if (value == null)
+                    // SQL Server allows 900 bytes per primary key, so use that as a baseline
+                    var combinedPrimaryKey = new StringBuilder(900 * table.PrimaryKeys.Count());
+
+                    // Look up primary key of T. Because we're going in the same order of fields every time,
+                    // we can assume that if two rows with the same primary key are in the list, they will collide
+                    foreach (PropertyInfo primaryKey in table.PrimaryKeys)
+                    {
+                        object value = primaryKey.GetValue(row);
+                        // Identity columns are allowed to be optional, so just skip the key if it doesn't exist
+                        if (value == null)
+                        {
+                            continue;
+                        }
+                        combinedPrimaryKey.Append(value.ToString());
+                    }
+                    // If we have already seen this unique primary key, skip this update
+                    if (uniqueUpdatedPrimaryKeys.Add(combinedPrimaryKey.ToString()))
                     {
-                        continue;
+                        // This is the first time we've seen this particular PK. Add this row to the upsert query.
+                        rowsToUpsert.Add(row);
                     }
-                    combinedPrimaryKey.Append(value.ToString());
                 }
-
-                // If we have already seen this unique primary key, skip this update
-                if (uniqueUpdatedPrimaryKeys.Add(combinedPrimaryKey.ToString()))
+                else
                 {
-                    // This is the first time we've seen this particular PK. Add this row to the upsert query.
+                    // ToDo: add check for duplicate primary keys once we find a way to get primary keys.
                     rowsToUpsert.Add(row);
                 }
             }
 
             rowData = JsonConvert.SerializeObject(rowsToUpsert, table.JsonSerializerSettings);
-            IEnumerable<string> columnNamesFromPOCO = typeof(T).GetProperties().Select(prop => prop.Name);
+            IEnumerable<string> columnNamesFromPOCO = this.GetColumnNamesFromPOCO();
```

Comment:
Instead of making this a member function it would be better to make `GetColumnNamesFromPOCO` into a static function and pass in the item to use if it's a JObject